src/DateRangePicker.ts

```diff
@@ -1,26 +1,423 @@
+  @property({ type: Number }) width = WIDTH;
+  @property({ type: Number }) height = HEIGHT;
+  @property({ type: Number }) sliderWidth = SLIDER_WIDTH;
+  @property({ type: Number }) tooltipWidth = TOOLTIP_WIDTH;
+  @property({ type: Number }) tooltipHeight = TOOLTIP_HEIGHT;
+  @property({ type: String }) dateFormat = DATE_FORMAT;
+  @property({ type: Object }) data?: HistogramInputData;
+  @property({ type: Object }) currentSlider?: SVGRectElement;
+
+  // these properties are intended to be private
+  @property({ type: Number }) _leftSliderX: number = 0;
+  @property({ type: Number }) _rightSliderX: number = 0;
+
+  // these properties don't need to be tracked for changes
+  _minDate: number = 0;
+  _maxDate: number = 0;
+  _dragOffset: number = 0;
+  _histWidth: number = 0;
+  _numBins: number = 0;
+  _binWidth: number = 0;
+  _histData: HistogramItem[] = [];
+
+  /* eslint-enable lines-between-class-members */
+
   static styles = css`
-    :host {
-      display: block;
-      padding: 25px;
-      color: var(--date-range-picker-text-color, #000);
+    #container {
+      margin: 0;
+      touch-action: none;
+      position: relative;
+    }
+    /****** histogram ********/
+    .bar:hover {
+      fill-opacity: 0.7;
+    }
+    #tooltip {
+      position: absolute;
+      display: none;
+      background: ${tooltipBackgroundColor};
+      color: ${tooltipTextColor};
+      text-align: center;
+      border-radius: 3px;
+      padding: 2px;
+      font-size: ${tooltipFontSize};
+      font-family: sans-serif;
+      touch-action: none;
+      pointer-events: none;
+    }
+    #tooltip:after {
+      content: '';
+      position: absolute;
+      margin-left: -5px;
+      top: 100%;
+      /* arrow */
+      border: 5px solid ${tooltipTextColor};
+      border-color: ${tooltipBackgroundColor} transparent transparent
+        transparent;
+    }
+    /****** slider ********/
+    .draggable:hover {
+      cursor: grab;
+    }
+    .dragging {
+      cursor: grabbing !important;
+    }
+    /****** inputs ********/
+    #inputs {
+      display: flex;
+      justify-content: center;
+    }
+    #inputs .dash {
+      position: relative;
+      bottom: 1px;
+    }
+    input {
+      width: ${inputWidth};
+      margin: 0 3px;
+      border: ${inputBorder};
+      border-radius: 2px !important;
+      text-align: center;
+      font-size: ${inputFontSize};
     }
   `;
 
-  @property({ type: String }) title = 'Hey there';
+  firstUpdated() {
+    this._leftSliderX = this.sliderWidth;
+    this._rightSliderX = this.width - this.sliderWidth;
+    this._minDate = dayjs(this.data?.minDate).valueOf();
+    this._maxDate = dayjs(this.data?.maxDate).valueOf();
+    this._histWidth = this.width - this.sliderWidth * 2;
+    this._numBins = this.data?.bins?.length ?? 1;
+    this._binWidth = this._histWidth / this._numBins;
+    this._histData = this.generateHistData();
+  }
+
+  generateHistData(): HistogramItem[] {
```

Comment:
This can be a getter

Reply:
Actually I did this deliberately since it takes a bit of work to calculate `this.generateHistData()` and I didn't want to repeat this work on every access of `this._histData` (ie on every render). In Ruby there is an easy way to memoize method results, but I'm not sure if it's worth doing that here inside of a getter. Any thoughts?

Reply:
It should only update if any of the reactive properties inside get updated like `this.data` and `this.height`. Were you seeing it recalculate a lot?

Reply:
Oh I think I see why you were seeing so many updates. In your `render()` method, you are calling the methods directly so it's executing it every update. ie:
```
${this.renderHistogram()} <-- ()
${this.renderSelectedRange()}
```
If you switch to using the getter approach and reference the template result, it will only update when dependent reactive properties get updated.. ie:
```
get histogramTemplate(): TemplateResult {
  const history = this.histData;
  return html`<div>... ${history} ...</div>`;
}

render(): TemplateResult {
  return html`${this.histogramTemplate}`; <-- now this will only update if any of the downstream dependencies get updated
}
```

Reply:
So I experimented with this a bit and accessing `this.histogramTemplate` calls the underlying getter method every time. LitElement still only updates the parts of the DOM that it needs to, but if I were to change `this.histData` into a getter, then it would have to iterate `this.data.bins` every update, rather than just on the first update.

Reply:
Interesting 🤔 Is that happening when you drag the endpoints around? I wouldn't think it'd have to recalculate this unless it needed to, but I might be missing something. Do you happen to have a branch with the reactive version? I'm curious to see it.